[21242] Move project required disk space counting to SQL

oliverguenther opened the pull request:
This adds a Project submodule `Project::Storage` that allows
counting required disk space for wikis, work packages, and repositories
in SQL for all projects and on-demand for a single project.

This removes all dependencies introduced from the ruby-based counting
(i.e., the acts_as_ountable gem).

Also changes the project information pane to show not  tooltips, but values for each module (when a storage count > 0 exists)
![bildschirmfoto 2015-08-17 um 12 52 47](https://cloud.githubusercontent.com/assets/459462/9303550/1ab770da-44df-11e5-8421-127ee47066b4.png)

Relevant work package: https://community.openproject.org/work_packages/21242

oliverguenther committed:
Move project required disk space counting to sql

This adds a model section `Project::Storage` that allows
counting required disk space for wikis, work packages, and repositories
in SQL for all projects and on-demand for a single project.

This removes all dependencies introduced from the ruby-based counting
(i.e., the acts_as_ountable gem).

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://oliverguenther-feature-projects-storage-refactor.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://oliverguenther-feature-projects-storage-refactor.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

NobodysNightmare commented on Gemfile:
```diff
@@ -45,7 +45,6 @@ gem 'warden-basic_auth', '~> 0.2.1'
 gem 'rails_autolink', '~> 1.1.6'
 gem "will_paginate", '~> 3.0'
 gem "acts_as_list", "~> 0.3.0"
-gem 'acts_as_countable', git: "https://github.com/finnlabs/acts_as_countable.git", ref: '2471265'
```
What about https://github.com/finnlabs/acts_as_countable then?

Shall we kill it after merging this PR?

oliverguenther replied:
I don't think it serves a broader purpose at the moment, so yes.

NobodysNightmare commented on app/controllers/admin_controller.rb:
```diff
@@ -51,8 +51,10 @@ def projects
+    @projects = Project.with_required_storage
+                .find :all,
+                      order: 'lft',
+                      conditions: c.conditions
```
More like a general question:

This will most likely collide with the Rails 4 PR (since `find :all, ...` is long deprecated and those occurences were replaced for the Rails 4 migration):
Should we convert such statements to `where(c.conditions).order('lft')` on the spot, because we are touching them anyway? Or should we leave them as-is?

oliverguenther replied:
I left this in there because [#3344](https://github.com/opf/openproject/pull/3344/files#diff-6d0eec853a873b030e87f5a21aecdf63R61) touches it again immediately afterwards, but otherwise I tend to refactor these whenever I find them.

NobodysNightmare commented on app/models/project/storage.rb:
```diff
@@ -0,0 +1,109 @@
+    # Count required disk storage for this project.
+    # Returns a hash of the form:
+    # 'total' => Total required disk space for this project
+    # 'modules' => Hash of localized keys and required space for this module
```
Just some nit picking about words: Wouldn't **localized** keys be keys that are already localized into the correct language? Don't you mean _localization keys_?

oliverguenther replied:
I did mean that, good catch. Will also note that `modules` contains only non-zero storage counts.

NobodysNightmare commented:
Since we talked a lot about performance and so on, I just looked at the execution plan for your queries using the MySQL workbench (this is the SQL being executed when looking at the project list):

![execution plan](https://cloud.githubusercontent.com/assets/453584/9304328/4e5ac894-44e7-11e5-9a95-dbf0767c1273.png)

Lets have a look at the red appearences:
- the full scan on the left is probably okay, because we are listing projects
- I am not sure whether the full index scans in the subqueries are too great, I think there is no way for MySQL to know, that only some parts of the subquery result will be needed later on, which is why full scans are performed (my assumption), I will check if I can come up with better SQL around here ^^
- I am not sure of the full table scan on the repos, might be due to my small local test instance, since there is an index over projects, which should definitely make sense to use on larger instances...

edit: Just to emphasize that: I did not find any real performance-wise problems (on my small-ish test instance), I just wanted to take a closer look at _possible_ problems...
Running blazingly fast for me ;)

NobodysNightmare commented on app/models/project.rb:
```diff
@@ -932,24 +930,6 @@ def archive!
-  end
-
-  def self.project_storage_expires_in
-    Setting.project_storage_cache_minutes.to_i.minutes
```
If I see this correctly we are no longer using this fancy cache setting, right?
Time to delete it again ^^

oliverguenther replied:
I have now [removed the projects-related setting](https://github.com/opf/openproject/pull/3368/files#diff-646ad6b10917e4385cbcc8c8524e24a2L281), the repository one is still in use.

oliverguenther commented:
Hmmm.. I'm no expert on the options that MySQL would provide for such queries. Having worked primarily with pg, I naturally wanted to use `DISTINCT ON (projects.id)` to aggregate both associations in a single query, but I couldn't figure out how to aggregate in a single compatible query while keeping both joins separated.

NobodysNightmare commented on app/views/projects/form/_required_storage.html.erb:
```diff
@@ -26,17 +26,19 @@ Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+               locals: { label: l(:label_total), value: storage['total'] } %>
+  <% end %>
 <% end %>
+</fieldset>
```
The `if ... end` and the `fieldset` are wrongfully interleaved:

At the moment you are rendering the `</fieldset>` (end-tag) alone, if there is no space required.

NobodysNightmare commented on spec/models/project/storage_spec.rb:
```diff
@@ -0,0 +1,114 @@
+      # be properly reloaded here in all cases.
+      project1.wiki.reload
+
+      expect(p1.work_package_required_space).to eq('2500')
```
You are native to PostgreSQL, aren't you?

I encountered PostgreSQL (or the AR-Adapter, don't know who of them is to blame) not being able to infer the integer type of such custom select columns.
This spec might fail for MySQL, because IIRC there the integer type of the result will correctly be recognized.

NobodysNightmare replied:
Regardless of the backend used: In a (dynamically) typed language, I would of course expect a result of the correct type. For the aggregated journals I hacked around the problem by overwriting the dynamic accessor by hand, doing the conversion using ARs conversion methods.

oliverguenther replied:
Yeah, mainly running pg locally and was waiting for travis this time. I'm usually running mysql on a VM, too, but that currently still runs other specs. I guess I'll have to run these more frequently [soon](https://github.com/basiszwo/openproject/commit/51a95fefc80615393a541d8548157c7619bb12c8)

But yeah, the workaround sounds reasonable (however the storage is not always computed).

NobodysNightmare replied:
Yeah, not sure how reasonable the workaround is in your case...

NobodysNightmare commented:
Regarding this SQL-performance-thing: I think one could improve the situation by employing a separate table for the required space that can be joined onto different tables.

Because in that case you would not need to rely on subqueries, but could use some JOINS, which would in turn be optimizable by the DBMS, because it would know when not all of those are needed.

Still: Lets start with the current solution and go further if we encounter issues. We should probably watch out for such issues, though.

oliverguenther commented:
One additional question remains: Now that we count the storage directly in SQL, for what actions should I  allow the repository to refresh its own count? I guess we could do that when someone browses the repository, as that's also the time when internal changesets are updated.

NobodysNightmare commented:
We only update the changesets when someone is browsing the repository?

Does that mean that mentioning a work package via commit will also not be noticed before then?

I only see these options for updates:
- post-commit/Push hooks
- whenever a specific method is called (and the last update is longer than x ago)
  - e.g. `count_required_storage`, this would be very close to your old implementation IIRC
- whenever a user looks at a specific site (and the last update is longer than x ago)
  - browse revisions
  - list projects
  - check config of a single project
  - other related pages
- on regular intervals (e.g. daily, weekly, etc)

Post-hooks are probably not always available, so we need at least one other option to be safe. In case there are hooks we should probably also disable the other update-method, to avoid useless updates when not many changes are pushed to a repo, but the available storage space (or the site hosting it) is checked regularly.

oliverguenther committed:
Workaround to PG adapter returning non-integer aggregated values

oliverguenther committed:
Extend autofetching of commits into update repository information

This commit extends the previous functionality of *autofetch commits*
into internal updating of repository information, which encompasses:

* Commit information (changesets)
* Disk space information

They are now both retrieved when browsing the repository and the latter
is fetched asynchronously whenever the last update is older than the
timeout setting.

This timeout setting is now user-configurable.

oliverguenther commented:
Post hooks are definitely the way to go, however out of scope of this PR and for 5.0, as we would have to find a to robustly manipulate all existing repositories to include these hooks.

My last commit adds a configurable option that encompasses both
-  the updating of changesets 
- the required disk space.

The former has always been retrieved synchronously on browsing the repository, despite this being a large burden for big repositories upon first call.
I'll keep this in mind when there's time to refactor this part, to make this part asynchronous.
The latter is now also initiated when browsing the repository, however remains asynchronous.

The following settings have been added to repositories.

![bildschirmfoto 2015-08-17 um 17 38 00](https://cloud.githubusercontent.com/assets/459462/9308881/bdba4a38-4506-11e5-809e-22a330f5a501.png)

If a user decides to deactivate that option, it has to fetch changesets using the /sys API (as before), and now also update the required storage via the same API.

That is, a user can already decide to opt-in to the hook-based configuration on his own.

oliverguenther commented:
@NobodysNightmare I've assigned this back to you. If you agree with the following changes (also elaborated above), this can be merged.
- workaround for non-integer aggregation in Postgres was added
- Repository information are updated when browsing the repository through the frontend to be consistent with the commit information. Storage cache time is now configurable through settings and may (optionally) be instead refreshed through a webhook in the sys API.

oliverguenther committed:
Use present instead of presence

oliverguenther committed:
Remove PG workaround required for Rails < 4

NobodysNightmare commented:
LGTM too :+1: 

But also waiting for travis to see if any unwanted things happened due to the Rails 4 merge...
